python/cuml/tests/test_umap.py

```diff
@@ -572,12 +572,12 @@ def test_fuzzy_simplicial_set(n_rows,
 @pytest.mark.parametrize('metric', ['l2', 'euclidean', 'sqeuclidean', 'l1',
                                     'manhattan', 'minkowski', 'chebyshev',
                                     'cosine', 'correlation', 'jaccard',
-                                    'hamming', 'canberra'])
+                                    'hamming', 'canberra', 'dice'])
```

Comment:
We can see if CI passes on this but I don't believe our dense pairwise distance API is currently able to compute Dice distance. This is a really trivial computation to do w/ any primitive that's capable of computing euclidean, though. It just requires the dot product and l1 norms:

```
value_t q_r_union = q_norm + r_norm;
value_t dice      = (2 * dot) / q_r_union;
bool both_empty   = q_r_union == 0;
return 1 - ((!both_empty * dice) + both_empty);
```

Let me know if this is something we should prioritize.

Reply:
This test passed for me locally before I put this PR up, but your point is well taken as the dense pairwise distance API doesn't support this currently. The existing machinery seems to handle some but not all of the behavior.

https://github.com/rapidsai/cuml/issues/5129#issuecomment-1397636438

Reply:
I'm honestly still scratching my head at how this could be working w/ dense inputs, which is making me wonder if we're getting a false positive in the passing test. The [FAISS brute-force API](https://github.com/rapidsai/raft/blob/branch-23.02/cpp/include/raft/spatial/knn/detail/common_faiss.h), which we're still using at the present time for non-L2 (and non-haversine) distances, doesn't support dice distance at all.